src/coreclr/jit/instr.cpp

```diff
@@ -795,11 +813,104 @@ CodeGen::OperandDesc CodeGen::genOperandDesc(GenTree* op)
+                    {
+                        assert(broadcastScalar->OperIs(GT_LCL_VAR));
+                        return OperandDesc(simdBaseType, broadcastScalar);
+                    }
+                }
+
+                case NI_SSE3_MoveAndDuplicate:
+                case NI_AVX2_BroadcastScalarToVector128:
+                case NI_AVX2_BroadcastScalarToVector256:
+                case NI_AVX512F_BroadcastScalarToVector512:
+                {
+                    assert(op->isContained());
+                    if (intrinsicId == NI_SSE3_MoveAndDuplicate)
+                    {
+                        assert(simdBaseType == TYP_DOUBLE);
+                    }
+                    // if broadcast node is contained, should mean that we have some forms like
+                    // broadcast -> CreateScalarUnsafe -> scalar.
+                    // if so, directly emit scalar.
+                    switch (simdBaseType)
+                    {
+                        case TYP_INT:
+                        case TYP_UINT:
+                        case TYP_LONG:
+                        case TYP_ULONG:
+                        {
+                            // a special case is when the operand of CreateScalarUnsafe is in integer type,
+                            // CreateScalarUnsafe node will be fold, so we directly match a pattern of
+                            // broadcast -> LCL_VAR(TYP_(U)INT)
+                            assert(hwintrinsic->Op(1)->OperIs(GT_LCL_VAR, GT_CNS_INT, GT_IND));
+                            GenTree* scalar = hwintrinsic->Op(1);
+                            assert(scalar->isContained());
+                            if (scalar->OperIs(GT_LCL_VAR, GT_IND))
+                            {
+                                // This handles the case:
+                                // BroadcastScalarToVector* -> LCL_VAR/IND.
+                                assert(scalar->isContained());
+                                return genOperandDesc(scalar);
+                            }
+                            else
+                            {
+                                // This handles the case:
+                                // BroadcastScalarToVector* -> CNS_INT.
+                                ssize_t        scalarValue = scalar->AsIntCon()->IconValue();
+                                UNATIVE_OFFSET cnum = emit->emitDataConst(&scalarValue, genTypeSize(simdBaseType),
+                                                                          genTypeSize(simdBaseType), simdBaseType);
+                                return OperandDesc(compiler->eeFindJitDataOffs(cnum));
+                            }
+                        }
+
+                        case TYP_FLOAT:
+                        case TYP_DOUBLE:
+                        {
+                            assert(hwintrinsic->isContained());
+                            assert(hwintrinsic->Op(1)->OperIs(GT_HWINTRINSIC, GT_IND));
```

Comment:
Similar here. I expect it would be good to just handle this recursively, especially since that also handles `CreateScalarUnsafe`.